### ISSUE TITLE
Add layer mechanism for composing dependencies (or effects in general?)

### ISSUE THREAD

**Issue by johnhungerford** [getkyo/kyo#153]
Given an effect `T < (S & Envs[A] & Envs[B] & Envs[C])`, it should be possible to construct a "layer" (terminology taken from ZIO) that will provide some or all of the dependencies `A`, `B`, and `C`, optionally requiring additional dependencies to satisfy these. You should be able to provide this layer to an effect to handle any of the `Envs` dependencies at once, returning an effect without the provided dependencies and (optionally) with additional dependencies.

I envision the signature as being:
```scala
type Layer[Sin, Sout]
```
where `Sout` are the effects being handled, and `Sin` are the effects needed to handle them. 

A layer could can be constructed and then provided to an effect as follows:

```scala
trait ElasticClient:
    def execute(request: ElasticRequest): List[ElasticResult] < (IOs & Aborts[ElasticError])

val elasticEffect: List[ElasticResult] < (IOs & Aborts[ElasticError] & Envs[ElasticClient]) =
    for {
        elasticClient <- Envs[ElasticClient].get
        results <- elasticClient.execute(someRequest)
    } yield results

// Call Envs.layer to construct a layer that will generate the dependency described by `Envs[ElasticClient]`
val elasticClientLayer: Layer[IOs & Tries & Envs[HttpClient], Envs[ElasticClient]] =
    Envs.layer {
        for {
            httpClient <- Envs[HttpClient].get
            elasticClient <- IOs {
                Tries.catching(ElasticClient(httpClient, ElasticClient.Config.default))
            }
        } yield elasticClient
    }

// Note that:
// 1) `Envs[ElasticClient]` is replaced by `Envs[HttpClient]`
// 2) The updated effect now has `Tries` in addition to `Aborts[ElasticError]`, due to the client constructor's effects
val updatedElasticEffect: List[ElasticResult] < (IOs & Tries & Aborts[ElasticError] & Envs[HttpClient]) =
    Layers.run(elasticClientLayer)(elasticEffect)
```

Layers should be composable in two ways:
1. Combine inputs and outputs
```scala
val layer1: Layer[A & B, X & Y]
val layer2: Layer[C, Z]

val composedLayer: Layer[A & B & C, X & Y & Z] = layer1 ++ layer2
```

2. Apply outputs of one layer to the inputs of another:
```scala
val layer1: Layer[A & B, C]
val layer2: Layer[C, X & Y]

val composedLayer: Layer[A & B, X & Y] = layer1 >>> layer2
```

Note that this mechanism could actually extend beyond the functionality of `ZLayer`, insofar as it could be used not only to satisfy `Envs` effects (dependencies), but could also be used to compose many different kinds of effect handlers.

For instance, say I have a domain error type `DomainError`, and there are a bunch of other error types I frequently have to convert to this error type. I can make a layer:

```scala
val elasticEffect: List[ElasticResult] < (IOs & Aborts[ElasticError] & Envs[ElasticClient]) = ??? // See above

val domainAbortsLayer: Layer[Aborts[DomainError], Aborts[ElasticError] & Aborts[IOError] & Aborts[HttpError]] =
    // Call Aborts.layer to construct a layer that handles error cases (provide an effect that will run on failure)
    Aborts.layer[ElasticError]((err: ElasticError) => Aborts[DomainError].fail(DomainError.fromElasticError(err))
        // Compose with other `Aborts` layers to handle different errors
        ++ Aborts.layer[IOError]((err: IOError) => Aborts[DomainError].fail(DomainError.fromIOError(err))
        ++ Aborts.layer[HttpError]((err: HttpError) => Aborts[HttpError].fail(DomainError.fromHttpError(err))

// Note that:
// 1) All the same effect transformations from the original example above are present, except:
// 2) `Aborts[ElasticError]` is replaced by `Aborts[DomainError]`
// 3) `Tries` is still present because it is not one of the `Sin` types of `domainAbortsLayer`
val updatedElasticEffect: List[ElasticResult] < (IOs & Tries & Aborts[ElasticError] & Envs[HttpClient]) =
    Layers.run(elasticClientLayer ++ domainAbortsLayer)(elasticEffect)
```

**Comment by johnhungerford** [comment]
I have an approach that seems to work well, if you're open to PRs

**Comment by deusaquilus** [comment]
@johnhungerford I’m definitely curious to see what you’ve got in mind!

**Comment by fwbrasil** [comment]
@johnhungerford Interesting! Yes, I'm definitely interesting in PRs :)

**Comment by johnhungerford** [comment]
Closed with #154
